src/general.py

```diff
@@ -155,7 +278,28 @@ def generate_sub_object(sub_file: Path, test_name: str) -> htcondor2.Submit:
         job["Requirements"] = f'TARGET.GLIDEIN_ResourceName == "$(ResourceName)"'
     else:
         job["Requirements"] = (
-            f'TARGET.GLIDEIN_ResourceName == "$(ResourceName)" && ' + req
+            f'(TARGET.GLIDEIN_ResourceName == "$(ResourceName)") && ({req})'
+        )
+
+    # add periodic removal statement
+    # job should be removed if it is in idle or running for more than 4 hours, if it ever
+    # goes on hold, or if it restarts more than 10 times
+    sec_in_4hr = 60 * 60 * 4
+    prdc_rm = job.get("periodic_remove")
+    if prdc_rm is None:
+        job["periodic_remove"] = (
+            f"(JobStatus == 1 && CurrentTime-EnteredCurrentStatus > {sec_in_4hr}) || "
+            + f"(JobStatus == 2 && CurrentTime-EnteredCurrentStatus > {sec_in_4hr}) || "
+            + "(JobStatus == 5) || "
+            + "(NumShadowStarts > 10)"
+        )
+    else:
+        job["periodic_remove"] = (
+            f"(JobStatus == 1 && CurrentTime-EnteredCurrentStatus > {sec_in_4hr}) || "
+            + f"(JobStatus == 2 && CurrentTime-EnteredCurrentStatus > {sec_in_4hr}) || "
+            + "(JobStatus == 5) || "
+            + "(NumShadowStarts > 10) || "
+            + f"({prdc_rm})"
```

Comment:
We should create our periodic remove expression once and store it in a variable. That way we don't have to worry about accidentally forgetting to update the expression in one place and not the other. (should probably do the same with the `requirements` expression.
Also, we should test the general expression (not the actual default numbers) for correct behavior. I can tell you how to modify things to test on Monday.